californium-core/src/main/java/org/eclipse/californium/core/server/ServerMessageDeliverer.java

```diff
@@ -154,19 +189,56 @@ private Resource findResource(final List<String> list) {
 		return current;
 	}
 
-	/* (non-Javadoc)
-	 * @see org.eclipse.californium.MessageDeliverer#deliverResponse(org.eclipse.californium.network.Exchange, org.eclipse.californium.coap.Response)
+	/**
+	 * Delivers an inbound CoAP response message to its corresponding request.
+	 * <p>
+	 * This method first invokes {@link #preDeliverResponse(Exchange, Response)}.
+	 * The response is considered <em>processed</em> if the
+	 * <em>preDeliverResponse</em> method returned {@code true}.
+	 * <p>
+* 	 * Otherwise, this method delivers the response to the corresponding request.
+	 * 
+	 * @param exchange
+	 *            The exchange containing the originating CoAP request.
+	 * @param response
+	 *            The inbound CoAP response message.
+	 * @throws NullPointerException if exchange or response are {@code null}.
+	 * @throws IllegalArgumentException if the exchange does not contain a request.
 	 */
 	@Override
-	public void deliverResponse(Exchange exchange, Response response) {
+	public final void deliverResponse(final Exchange exchange, final Response response) {
 		if (response == null) {
 			throw new NullPointerException("Response must not be null");
 		} else if (exchange == null) {
 			throw new NullPointerException("Exchange must not be null");
 		} else if (exchange.getRequest() == null) {
 			throw new IllegalArgumentException("Exchange does not contain request");
 		} else {
-			exchange.getRequest().setResponse(response);
+			boolean processed = preDeliverResponse(exchange, response);
+			if (!processed && exchange.getRequest().getResponse() == null) {
+				exchange.getRequest().setResponse(response);
```

Comment:
`getResponse() == null`

I'm not sure, if this additional test is valid for notifies. According `Request.setResponse(Response)`, the intention related to notifies seems not to mach that check. (Currently this seem just to be broken, that notify responses are delivered to the request, but this depends more on the shallow clone of the observe request, when notifies are processed.)

Reply:
I am not sure I understand your point. Are you saying that for notifications (which are responses to an original request), the response is not (or should not be) delivered to the request?

Reply:
The PR added an new check in line 217, the `getResponse() == null`. So the first impression of that would be, that **NO** newer `Responses` will be set, because a old `Response` will cause that new check to fail.
Currently in 2.0.x, notifies are processed with "shallow clones" of the original `Request`, which also removes a already received response, and therefore the new check didn't fail.
But the downside of the current solution in 2.0.x is, that observe/notify using the `Request.getResponse()` on the original request is broken  and so the cf-plugtest-checker fails also.

FMPOV, I would prefer **not** to manifest that "current 2.0.x way" of processing notifies by adding this check in line 217!

Reply:
Now I understand :-) I agree, we should then simply remove the `getResponse() == null` check and rely on the returned boolean only. WDYT?

Reply:
That was exactly what I want.

Reply:
what about the `deliverRequest` method? Should we omit the check for the existence of the response there as well? IMHO we do not have the same potential problem there as we have with notifications, do we? However, for consistency reasons we might be inclined to check for the returned boolean only as well. WDYT?

Reply:
I just don't know, why this check was added.  

Without transparent blockwise, I would guess, that `UdpMatcher.receiveRequest()` will either create a new `Exchange` (therefore without `Response`), or mark the `Request` as duplicate, but then the `ReliabilityLayer` will not forward it to the other layers and it would therefore not be passed to `ServerMessageDeliverer`.

So the check in line 89 would only check, what ever happens in `preDeliverRequest`.
Therefore I don't see any conflict (and also no benefit) in that check.

Reply:
> I just don't know, why this check was added.

It is just for making sure that the request has not been processed already. This is just a precaution to double-check that the `preDeliverRequest` method has not already created a response albeit having returned `false`.

Reply:
FMPOV remove that "precaution to double-check". It's easy to do it right in preDeliverRequest, and, if a not expected result is delivered, that would cause the implementer to do it proper.